### ISSUE TITLE
mixing depfile absolute paths with manifest relative paths breaks rebuilds

### ISSUE THREAD

**Issue by bradking** [ninja-build/ninja#1251]
The ninja manual warns against such mixture, but it has become a serious limitation.

The problem is that people want their build systems to invoke the compiler with an absolute path to the source file.  This produces better output for IDEs to match error messages and refer back to the original files.  It also may affect the paths recorded in debug information.  However, this conflicts with ninja's preference for relative paths in the build manifest.

Here is a script demonstrating the issue: [ninja-bug.bash.txt](https://github.com/ninja-build/ninja/files/799752/ninja-bug.bash.txt)

**Comment by bradking** [comment]
Cc: @nico @chaoren

Over in CMake, there is relevant discussion in [issue 16675](https://gitlab.kitware.com/cmake/cmake/issues/16675), [issue 13894](https://gitlab.kitware.com/cmake/cmake/issues/13894), and [MR 520](https://gitlab.kitware.com/cmake/cmake/merge_requests/520).

**Comment by zlolik** [comment]
As workaround you can set object output directory explicitly, even "." or "..", but absolute path works also.
[ninjatst.bash.txt](https://github.com/ninja-build/ninja/files/822532/ninjatst.bash.txt)

**Comment by bradking** [comment]
@zlolik IIUC you suggest placing absolute paths in the build manifest (`build.ninja`) graph nodes.  Certainly that works (and is what CMake does in out-of-source builds), but my point here is that Ninja discourages use of absolute paths and yet IDEs require them in output.

It is also hypothetically possible that a compiler given `-c foo.c` could generate an absolute path in the depfile anyway.  If that happens then Ninja will not recognize the dependency due to the same problem reported here.

Fixing this will require Ninja to be able to recognize when a relative path and absolute path are the same file and have only one graph node internally instead of two.  Doing this only for depfile processing may be sufficient, as it is reasonable to ask the `build.ninja` generators to name paths consistently within the build manifest.

**Comment by zlolik** [comment]
@bradking First of all, I am agree with you that we have an issue with ninja and I only suggest workaround.
I think there are 3 linked issues:
1. Logic with input absolute/relative path, more details in #1152 .
2. Logic to force output absolute/relative path. The problem with compilers - the only way to said to compiler that you want to have absolute path in output is use absolute path in command line. Not ninja issue but ours.
3. Logic to parse depfile absolute/relative path. This issue.

I use 2 workarounds:
1. Use prefixes `$p` and `$o` in path of `build.ninja` generator to solve first 2 problems. In my case `$p` and `$o` are not absolute but relative from current dir where `build.ninja` generator was run.
2. Use some script to convert depfile to ninja compatible path. Some compilers even need separate call to generate depfile. But there is no absolute path problem while.

**Comment by bradking** [comment]
For reference, I've started a [thread](https://groups.google.com/forum/#!topic/ninja-build/yJvs7u0n2iA) on the `ninja-build` mailing list with a proposed solution.

**Comment by chfast** [comment]
Another related issues about debug and coverage information: using commands with relative paths creates also debug information files with relative paths. The problem is the debug files are usually placed next to the object files, not the source files, so the relative paths there reference non-existing files.

**Comment by thughes** [comment]
Any update on this? There haven't been any responses on the mailing list thread...

**Comment by sebknzl** [comment]
Hello everyone,

> It is also hypothetically possible that a compiler given `-c foo.c` could generate an absolute path in the depfile anyway.

IAR v8's dependency files contain absolute paths only.

We've hit this problem when doing out-of-source builds with CMake/Ninja/IAR: Generated files which end up in CMake's binary dir are relative in the Ninja manifest, the deps output by IAR are absolute.

To me, this is clearly a Ninja-issue: It shouldn't have two nodes for the same file, no matter what.

I've read @bradking 's [suggestion](https://groups.google.com/forum/#!topic/ninja-build/yJvs7u0n2iA), but I don't really understand the part about symbolic links and why `getcwd()` can't be used. Would you mind elaborating?

**Comment by bradking** [comment]
> I don't really understand the part about symbolic links and why `getcwd()` can't be used

@sebknzl the problem is that the absolute paths written to `build.ninja` by the generator may contain symbolic links that are intentionally not resolved for one reason or another.  `getcwd()` typically returns a path with symlinks resolved.  If Ninja were to interpret a relative path in the build manifest or a depfile relative to `getcwd()` then the absolute path it constructs may not match what the build manifest generator used and we'd still end up with two nodes.

The `ninja_workdir` I propose in the thread you linked is basically telling Ninja what the generator used as the base for relative paths.  In practice  `realpath(ninja_workdir) == realpath(getcwd())` should always be true, but that does not mean `ninja_workdir == getcwd()`.  Another alternative would be for Ninja to `realpath` everything before creating nodes, but that would mean `$in` would not expand to what the generator intended.

**Comment by bradking** [comment]
@jhasse please see this issue and [my proposal](https://groups.google.com/forum/#!topic/ninja-build/yJvs7u0n2iA) to address it.

**Comment by ClausKlein** [comment]
Why not generate a **_build.ninja_** with absolute paths controlled by an **CMAKE_NINJA_GENERATOR_OPTION_** ?

i.e. like  this, which works:

```ninja
Claus-MBP:build clausklein$ cat build.ninja 

# project dir
p = /Users/clausklein/Downloads
# object dir
o = /Users/clausklein/Downloads/.ninjatst.build

rule cp
  command = cp $in $out

rule cc
  depfile = $out.d
  deps = gcc
  command = cc -I$o -c $in -o $out -MD -MT $out -MF $out.d

# All absolute paths works
build all: phony $o/foo.o

# All absolute paths works
build $o/foo.h: cp $p/foo.h.in
  IN = $p/foo.h.in

# All absolute paths works
build $o/foo.o: cc $p/foo.c || $o/foo.h
  IN = "$p/foo.c"

default all

Claus-MBP:build clausklein$ 
```
generated with [ninjatst.bash.txt](https://github.com/ninja-build/ninja/files/4320852/ninjatst.bash.txt)

**Comment by bradking** [comment]
@ClausKlein we did try teaching CMake to use absolute paths.  See the discussions I linked in https://github.com/ninja-build/ninja/issues/1251#issuecomment-282322776.  I'd rather not add an option that says "do things in a different way that fixes some cases and breaks others".

We really need the Ninja feature in the proposal I linked in https://github.com/ninja-build/ninja/issues/1251#issuecomment-487618865 to fix this properly.

**Comment by ClausKlein** [comment]
OK, fine

Then would this fix for mesonbuild system available too ;-)

**Comment by jhasse** [comment]
Let's say Ninja would use `ninja_workdir` for matching depfile information and we have the situation where `ninja_workdir != getcwd()` because of a symlink. What if the compiler now prints absolute paths by using `getcwd()`?

**Comment by bradking** [comment]
The compiler should not use `getcwd()` for anything unless those things were passed in as relative paths, in which case the buildsystem generator likely needs to take responsibility.

Or we can consider trying to normalize w.r.t. both `ninja_workdir` and `getcwd()`.

**Comment by jhasse** [comment]
Well, the compiler *should* also not return an absolute path in the depfile when given relative paths ;)

> @ClausKlein we did try teaching CMake to use absolute paths. See the discussions I linked in #1251 (comment). I'd rather not add an option that says "do things in a different way that fixes some cases and breaks others".

What exactly broke when you used absolute paths as suggested by @ClausKlein?

**Comment by bradking** [comment]
If the compiler is given absolute paths then the depfile it generates has absolute paths, and those are not reconciled with relative paths to headers in the build manifest, e.g. for generated headers.  See the `ninja-bug.bash.txt` link in my first post at the top of this issue.  It shows a step-by-step example of what happens and is expressed purely in terms of Ninja and no CMake.

**Comment by jhasse** [comment]
I've read that, but why not use https://github.com/ninja-build/ninja/issues/1251#issuecomment-284533599? Simply because Ninja "discourages" absolute paths? If so, this is about a documentation change?

**Comment by bradking** [comment]
`build.ninja` files are much smaller without repeating absolute paths everywhere, and that makes them faster.  They also look nicer, are easier to debug, and are more appropriate for tab-completion of build target names in shells.  IMO it is a good recommendation for Ninja to make.  Ninja just needs a bit more information to reconcile this case, and that is what I propose.

**Comment by ClausKlein** [comment]
@bradking relative paths may not always short:

```ninja
build jsoncpp@sha/src_lib_json_json_reader.cpp.o: cpp_COMPILER ../../../../../Users/clausklein/Workspace/cpp/jsoncpp/src/lib_json/json_reader.cpp
 DEPFILE = jsoncpp@sha/src_lib_json_json_reader.cpp.o.d
 ARGS = -Ijsoncpp@sha -I. -I../../../../../Users/clausklein/Workspace/cpp/jsoncpp -I../../../../../Users/clausklein/Workspace/cpp/jsoncpp/include -I../../../../../Users/clausklein/Workspace/cpp/jsoncpp/dir1 -I../../../../../Users/clausklein/Workspace/cpp/jsoncpp/dir2 -fdiagnostics-color=always -pipe -Wall -Winvalid-pch -Wnon-virtual-dtor -std=c++17 -O3
```
my example https://github.com/ninja-build/ninja/issues/1251#issuecomment-597877775 looks nicer!

**Comment by bradking** [comment]
CMake's generators have a rule about not using `../` sequences that leave the build tree.  In fully out-of-source builds we do use absolute paths to the source tree and relative paths within the build tree.  The latter never start in `../` because they are under the build tree.

**Comment by ClausKlein** [comment]
I know, it was generated with **_mesonbuild_**. ;-)

But than, why not deny the generation into source tree?

**Comment by bradking** [comment]
We can debate the merits of relative v. absolute paths forever.  Ninja should not be opinionated about that.  The problem here is reconciling cases where both are used within a single build.  The buildsystem generator cannot always control what the tools will do with paths.  Ninja should support mixed absolute/relative paths to the same file and can do so with a simple approach I've outlined in my proposal.  For the motivating use case the changes are isolated to depfile parsing.

**Comment by ClausKlein** [comment]
> Ninja should support mixed absolute/relative paths to the same file

I agree!

**Comment by jhasse** [comment]
Thanks for all the explanations, I'm starting to understand.

What if Ninja would (upon encountering an absolute path in the depfile) realpath both cwd and the path it found? Then it would remove the realpathed cwd from the realpathed depfile-path to get the relative path to match against the generated header. Would that work?

**Comment by bradking** [comment]
`getcwd()` is always a realpath already so that step isn't necessary.  If Ninja were to `realpath()` the depfile-provided absolute path and check for a cwd prefix that could work.  However, `realpath()` is a possibly-expensive syscall while my proposed `ninja_workdir` can reconcile paths purely in memory via string processing.

**Comment by bradking** [comment]
Mixing in the other direction should also be allowed: an absolute path in the build manifest but a relative path in the depfile.  I don't think Ninja should have to `realpath()` every absolute path while parsing `build.ninja`.  That would probably be slow.

**Comment by jhasse** [comment]
We could use the result to determine the workdir. E.g.:

getcwd(): `/var/xyz/build`
depfile entry: `/home/foo/build/bar.h`
realpath(depfile entry): `/var/xyz/build/bar.h`
relative realpath: `bar.h`
remove that from the depfile entry: `/home/foo/build` -> ninja_workdir

This way we only need to call realpath once.

**Comment by bradking** [comment]
I do not think deriving `ninja_workdir` automatically is possible in general.  How do we know how many logical components of the original symlink-containing logical path correspond to the workdir?  We might be able to do it by taking one component at a time until its realpath matches the cwd, but that feels hacky and error prone.  It feels risky to trust the first symlink that happens to look like this.  Also we'd have to run realpath() on every absolute path we encounter until one resolves this way.

The buildsystem generator knows exactly what logical path to the workdir it used when generating absolute paths in build statements and command lines in `build.ninja`.  It is therefore the most reliable source of information for the workdir path.  We just need a way to communicate this information to Ninja, hence my proposed `ninja_workdir` binding.

**Comment by jhasse** [comment]
I'm not sure I understand, sorry. Can you give an example where my detection logic wouldn't work?

**Comment by bradking** [comment]
It could break if the buildsystem generator puts a symlink inside the build tree (pointing elsewhere in the build tree or outside altogether):

getcwd(): `/var/xyz/build`
depfile entry: `/home/foo/build/subdir-link/bar.h`
realpath(depfile entry): `/var/xyz/build/subdir-real/bar.h`
relative realpath: `subdir-real/bar.h`
remove that from the depfile entry: ???

CMake does not do this but in general a buildsystem generator could.

Even if we are willing to say that is not supported, the detection could still face:

getcwd(): `/var/xyz/build`
1000 depfile entries: `/home/foo/ext-lib/bar{1,2,...}.h`
1000 realpath(depfile entries): `/var/xyz/ext-lib/bar{1,2,...}.h`
1000 relative realpaths: none under cwd

Now imagine that appears in 1000 sources' depfiles and there is never a hit.  That is 1 million realpath() calls with no match for this heuristic.  And that repeats every time a source is rebuilt and its depfile re-parsed.

**Comment by jhasse** [comment]
symlink inside the build tree -> yes, we can't really support that, it will result in all kinds of other issues anyway (without realpath-ing all over the place).

The second example sounds rather artificial. Normally depfiles start with the local dependencies, not with system headers.

If it really turns out to be a performance problem, we could cache the workdir in `.ninja_deps`?

**Comment by bradking** [comment]
In an out-of-source build very few depfile entries will even point at the build tree.  This will be very common:

getcwd(): `/var/xyz/build`
1000 depfile entries: /home/foo/project-source/bar{1,2,...}.h
1000 realpath(depfile entries): /var/xyz/project-source/bar{1,2,...}.h
1000 relative realpaths: none under cwd

Ninja will have to call `realpath()` on all of those just in case, or maintain some kind of table of results to avoid the actual syscalls.  It cannot know ahead of time when it will encounter a symlink path that resolves to the cwd, and does not even know *if* it will find one.

**Comment by jhasse** [comment]
Given that mixing absolute and realtive paths in the depfile currently doesn't work, I don't see how that would be "very common"??

**Comment by bradking** [comment]
depfiles of the form in my example are already very common even when not using symlinks or mixed paths.  I have them in most of my build trees.  Ninja can't know ahead of time whether the heuristic is needed so it will have to block on `realpath()` syscalls just in case.

**Comment by jhasse** [comment]
Maybe I misunderstood your example. Do you mean one depfile with 3000 entries or were those 3 seperate examples for depfiles with 1000 entries each?

**Comment by bradking** [comment]
I meant 1000 depfile entries spread across any number of depfiles. It could be one.  It could be many.  There could be 10000 entries across a project.  Their paths can point anywhere on the filesystem and may or may not ever actually point into the build tree.

**Comment by jhasse** [comment]
Can you give an example `CMakeLists.txt` that results in such a mixture of paths (doesn't have to be 1000, just one of each)?

**Comment by bradking** [comment]
It's hard to show in a `CMakeLists.txt` file because CMake currently takes care to avoid mixing paths in most cases.  However, there are issues in CMake's issue tracker about problems that cannot be fixed without Ninja fixing this first.  See for example [CMake Issue 13894](https://gitlab.kitware.com/cmake/cmake/-/issues/13894).  Users want to be able to give absolute paths to the compiler even for sources (and perhaps include directories) in the build tree, or when doing in-source builds.  This works fine with CMake's Makefile generator.  If we do that with Ninja then the depfiles come back with absolute paths and they don't match the relative paths in `build.ninja`.  Maybe we could switch to absolute paths for everything, but as discussed above here we shouldn't have to do that.

**Comment by ClausKlein** [comment]
perhaps you may try this for tests https://github.com/ClausKlein/build-performance/commit/8013836486e3a459474eb374f6c3da5e20983443

**Comment by jhasse** [comment]
> This works fine with CMake's Makefile generator.

How does Make handle the symlink issue?

**Comment by bradking** [comment]
> How does Make handle the symlink issue?

The Makefile generator is not currently using depfiles and does its own approximate scanning.  That will likely be updated as part of adding C++20 modules support to the Makefile generator, at which point the interpretation of depfile paths will have to be reconciled similarly to what is proposed here.

**Comment by bradking** [comment]
One option is to skip `ninja_workdir` for now and at least teach Ninja to reconcile the mixed paths assuming that `getcwd()` is the top.  That way at least the mixed paths will work in the common no-symlink case.  Then `ninja_workdir` can be added to support the more complex cases later.

**Comment by GrandChris** [comment]
Pleas someone fix this. I'm so annoyed because of this.

 I'm using ARM-GCC with CMake and VSCode. When I use an in-source build, the paths are relative to the "build" directory (instead of the root directory), which causes VSCode to no longer be able to resolve the paths, which means, you can't just click the paths in the command line window, instead, you have to locate each file yourself. When I use a CMake out-of-tree build, it breaks the "IntelliSense" of Eclipse and makes problems with CMake subprojects. My solution for now is to use MakeFiles, because it uses absolute Paths per default. But Ninja is better, it produces better readable files (rules.ninja...) and the build output on the command line is better. Make has problems with outputting warnings on the command line in a multithreaded context, means, multiple warnings are interleaved. Which causes me to build single threaded, which is again, super annoying.

**Comment by GrandChris** [comment]
Thank you ClausKlein, but the workaround is an out of source build, but I need the files inside my tree, because it is referenced by other scripts. Also it is a non-standard build requirement.

My solution for now is, to change CMake to just always use absolute paths. I don't understand why there isn't an option for this. However, if somebody else needs this, here is the patch of CMake v3.18.1:
"
@@ -241,7 +241,7 @@ void cmNinjaTargetGenerator::AddIncludeFlags(std::string& languageFlags,
   // Add include directory flags.
   std::string includeFlags = this->LocalGenerator->GetIncludeFlags(
     includes, this->GeneratorTarget, language,
-    language == "RC", // full include paths for RC needed by cmcldeps
+    true, // full include paths for RC needed by cmcldeps
     false, config);
   if (this->GetGlobalGenerator()->IsGCCOnWindows()) {
     std::replace(includeFlags.begin(), includeFlags.end(), '\\', '/');
@@ -1133,8 +1133,7 @@ void cmNinjaTargetGenerator::WriteObjectBuildStatement(
   const std::string& fileConfig, bool firstForConfig)
 {
   std::string const language = source->GetLanguage();
-  std::string const sourceFileName =
-    language == "RC" ? source->GetFullPath() : this->GetSourceFilePath(source);
+  std::string const sourceFileName = source->GetFullPath();
   std::string const objectDir = this->ConvertToNinjaPath(
     cmStrCat(this->GeneratorTarget->GetSupportDirectory(),
              this->GetGlobalGenerator()->ConfigDirectory(config)));
"

**Comment by ClausKlein** [comment]
**Attention**, there was a patch for cmake, but see https://github.com/ninja-build/ninja/issues/1251

**Comment by GrandChris** [comment]
In my last comment, I showed a patch for CMake to use absolute paths instead of relative paths. Turns out this is a bad idea, because it breaks custom CMake header dependencies with Ninja (for example a generated header file is dependent on a .json file). Guess this problem derives from this issue here, which I was too stupid to understand.

**Comment by GrandChris** [comment]
If someone is interested about it, I now use Ninja with relative paths. I have created an extension for VSCode to be able to click relative paths inside the integrated terminal. [extension](https://marketplace.visualstudio.com/items?itemName=GrandChris.just-let-me-click-this-freaking-paths)

**Comment by weliveindetail** [comment]
@GrandChris Great! Any chance to get it to work in the `Output` tab too?

**Comment by GrandChris** [comment]
@weliveindetail Feel free to create a new issue on https://github.com/GrandChris/TerminalRelativePath/issues and describe your use-case in detail. But the output tab in VSCode behaves different to the integrated terminal and I don't think there is an api to extend the behavior in this regard.

**Comment by weliveindetail** [comment]
Ok thanks. Eventually, clickable relative paths should probably be solved in VSCode or the CMake extension. Will ping on their side.

**Comment by bradking** [comment]
For reference, this came up again in [CMake Issue 21865](https://gitlab.kitware.com/cmake/cmake/-/issues/21865#note_910591).  It is another example of a case where it is not easy (possible?) to make depfile-discovered dependencies use the exact same path representation as `build.ninja`.

#1917 adds the above-proposed `ninja_workdir` binding as a feature useful on its own.  Once that is merged, I can follow up with a fix for this issue.

**Comment by bradking** [comment]
@jhasse I'd like to come up with an acceptable solution to get this fixed upstream.  `ninja_workdir` solves the problem.  I've argued my case above.  What other alternative do you suggest?

**Comment by jhasse** [comment]
As you've pointed the issue (mixing depfile absolute paths with manifest relative paths breaks rebuilds) is by design. I think we should create more specific issues, because we have touched on several issues in the discussion (e.g. symlinking the build dir).

For the two usecases in the initial comment my alternatives would be:

> This produces better output for IDEs to match error messages and refer back to the original files.

Either fix the IDE, or pass [`/FC`](https://docs.microsoft.com/de-de/cpp/build/reference/fc-full-path-of-source-code-file-in-diagnostics) or [`-fabs-path-diagnostics`](https://reviews.llvm.org/D23816) to the compiler.

>  It also may affect the paths recorded in debug information.

This should also be fixed by the compiler. I would suggest looking at `-fdebug-prefix-map`, but I'm not sure what the issue is exactly.

> Here is a script demonstrating the issue: ninja-bug.bash.txt

Not really, it demonstrates Ninja's current behavior. The two issues

1. "my IDE can't match the compiler's output to the correct file"
2. "I need absolute paths in the debug information"

need further explanation. Otherwise this looks like an [XY problem](https://en.wikipedia.org/wiki/XY_problem) to me (X being "mixing absolute and relative paths" and Y being either 1. or 2.).

**Comment by bradking** [comment]
I really don't think the solution should be "work around this in all other tooling".

#1924 demonstrates a complete working fix with tests, based on the `ninja_workdir` approach.  A generator using this can solve all of the above problems.

**Comment by jhasse** [comment]
Me neither, because I don't see those as workarounds.

**Comment by ClausKlein** [comment]
This are NOT GNU gcc options!

> Am 05.03.2021 um 16:58 schrieb Jan Niklas Hasse <notifications@github.com>:
> 
> Either fix the IDE, or pass /FC <https://docs.microsoft.com/de-de/cpp/build/reference/fc-full-path-of-source-code-file-in-diagnostics> or -fabs-path-diagnostics <https://reviews.llvm.org/D23816> to the compiler.
> 
> It also may affect the paths recorded in debug information.
>

**Comment by bradking** [comment]
>> Here is a script demonstrating the issue: ninja-bug.bash.txt
>
> Not really, it demonstrates Ninja's current behavior.

My post you quoted acknowledges that this is currently the documented behavior.
The purpose of this issue is to call the **design** into question by
demonstrating a class of use cases that it does not serve.

> The two issues ... (1) ... (2) ... need further explanation.

Those are only two examples of problems that are made easier to solve by
passing absolute paths to tooling.  Absolute paths are not unreasonable
just because other solutions are available in specific instances of such
problems (using specific features of tooling available from some vendors).
Absolute paths help solve a large class of these problems without updates
to the rest of the ecosystem.  As I [posted previously](https://github.com/ninja-build/ninja/issues/1251#issuecomment-598307127):

> We can debate the merits of relative v. absolute paths forever. Ninja should
> not be opinionated about that. The problem here is reconciling cases where
> both are used within a single build.  The buildsystem generator cannot always
> control what the tools will do with paths.

In complex real-world build systems, there can be an arbitrary number of layers
of scripts, wrappers, tools, etc., (from all kinds of vendors) between the
paths written in `build.ninja` by the build file generator and the actual paths
encountered by Ninja in a depfile.  Sometimes a given file can be referenced by
a depfile from a tool that was not directly given a path from `build.ninja`.
In order for depfiles to work under Ninja's current design, the entire tooling
stack has to be told exactly how the `build.ninja` generator chooses to express
those paths on build statements, and have features preserve them or convert
them in the same way.  This is not tractable in general.

Ninja can resolve this problem with a small design change: reconciling an
absolute path in a depfile with a relative path in `build.ninja` (and vice
versa).  I'm not asking to reconcile arbitrary paths to the same file, only to
recognize that `/path/to/some/file` in a depfile is the same as `some/file` in
the build file `/path/to/build.ninja`.

There is precedent for such design updates.  When Ninja first started, the hope
was that paths would always be encountered by `ninja` in exactly the same
byte-for-byte representation that the build file generator wrote in
`build.ninja`.  Run `git blame -M -C -w -- src/util.cc` and look at the history
of `CanonicalizePath` to see a series of cases that were encountered where that
design philosophy had to be relaxed.  This is another class of such cases.

**Comment by jhasse** [comment]
The thing is that I disagree with

> Ninja should not be opinionated about that.

in the first place.

**Comment by jhasse** [comment]
> This is not tractable in general.

So what is the specific issue? Sorry I'm a bit confused.

**Comment by bradking** [comment]
>  I disagree with "Ninja should not be opinionated about that" in the first place....So what is the specific issue?

This is not about any one specific issue, but about an *entire class* of problems that are *much easier* to solve with absolute paths.  It shouldn't matter *why* developers want to pass absolute paths to to their tools, only that they want to use absolute paths for whatever reasons apply to their particular case.  Build systems exist to serve their users, not to dictate policy.  Ninja is supposed to be an assembly language for build systems, and yet has no way to express build systems mixed-style paths.

Ninja's policy for upstream changes is that problems should be solved by build file generators, and not Ninja upstream, if they can be solved *without loss of efficiency*.  In this case, the only way to provide general-purpose `ninja_workdir`-style reconciliation of depfile paths from the build file generator is to inject after every command that generates a depfile an extra command to revise paths in the depfile to match the build file before ninja loads it.  That extra command will have its own runtime cost that is far above what #1924 does (a small number extra in-memory lookups inside the `ninja` process).  Therefore the general-purpose reconciliation is more efficient to provide in Ninja upstream, and so is a candidate for inclusion.

**Comment by comex** [comment]
Even if you do use absolute paths for everything, rather than mixing absolute and relative paths, there's an annoying ergonomics issue.  If the user asks `ninja` to regenerate a specific output file rather than a phony target, e.g., `ninja out/my_executable` or `ninja out/foo/bar.o`, they have to remember to pass the absolute path, or ninja will fail with a confusing "unknown target" error.  The opposite problem also exists: the user might ask ninja to rebuild an absolute path when the build.ninja file uses relative paths.

In one Ninja generator tool I maintain, I work around this by having every single `build` statement include both the relative path and its corresponding absolute path as outputs.  It works, but it really feels like a hack to work around a limitation in Ninja.

Admittedly, this isn't exactly the same issue as depfiles, but it suggests Ninja should be more aware of absolute vs. relative path equivalences in general.

**Comment by jhasse** [comment]
@comex Thanks, that's a valid concern. Can you open a new issue for that?

>  It shouldn't matter why developers want to pass absolute paths to to their tools [...]

I disagree.

> In this case, the only way to provide general-purpose `ninja_workdir`-style reconciliation of depfile paths [...]

Why do you need the `ninja_workdir`-style reconciliation in the first place though?

**Comment by jhasse** [comment]
> [...] absolute paths generated in depfiles by tooling they do not control.

Is there an example for such tooling?

**Comment by ClausKlein** [comment]
Yes, that is true!

And with absolute paths, cmake is able to eliminate duplicate compiler include paths .
This reduce the build time again.

> Am 09.03.2021 um 19:20 schrieb Brad King <notifications@github.com>:
> 
> I don't think it is the place of a build tool to evangelize relative paths.

**Comment by wackinger** [comment]
I setup the CMake issue regarding a problem we are facing with Qt's Quick compiler generated cpp files and and enabled unity build. My local investigations showed that ninja is by far fastest build system, but in this constellation it is unreliable. Now I'm facing the problem how I can find a workaround or justify the build time increase.

**Comment by KyleFromKitware** [comment]
CMake just encountered [an issue](https://gitlab.kitware.com/cmake/cmake/-/issues/21977) with circular dependencies that was, until recently, masked by this Ninja issue. If this issue had been fixed, we would have discovered the CMake issue a lot sooner.

**Comment by kobalicek** [comment]
I'm for merging the mentioned PR - https://github.com/ninja-build/ninja/pull/1924

It solves the problem many users have and it doesn't hurt ninja in any way. I really don't understand why it's so hard to fix ninja issues.

**Comment by JBakamovic** [comment]
I'm experiencing problems with my dev environment for years now exactly because of this issue. It's not a deal-breaker per se, I still very much prefer using `ninja` because of better development turn-around times, but I must say it's very annoying because, whenever `compile_commands.json` is regenerated (on big project every day), I must not forget to re-apply the `sed` voodoo which will fix the relative paths in it. If I do, and I do, tooling will start to fail miserably because it will not be able to find appropriate paths anymore and thus basically any operation that comes with it will stop working reliably or will stop working at all (indexing, jumping to definition, finding all references, semantic syntax highlighting, etc.). This is very disruptive and something nobody wants to keep in their mind while doing the development.

Tooling is based on LLVM/Clang infrastructure (libclang) and uses `compile_commands.json` to extract relevant information about paths, directories, build flags etc. Given that there are many dev tools nowadays which are built on top of this combination too, I'm sure that they are:
1. Either suffering from the relative path problems and have no workarounds implemented for them
2. Or they've implemented some sort of heuristics to workaround the problem

This is either bad for the end users (devs using the tooling which do not have workarounds) or it's bad for the tooling devs because they have to maintain it and make sure it does not break in future releases. I think that giving a chance to implant this behavior in `ninja` will let alone make many devs more happier and more productive.

**Comment by jhasse** [comment]
> whenever compile_commands.json is regenerated

by CMake or the compdb tool?

**Comment by JBakamovic** [comment]
`cmake ../ -GNinja -DCMAKE_EXPORT_COMPILE_COMMANDS=ON`

**Comment by duanyutong** [comment]
> `cmake ../ -GNinja -DCMAKE_EXPORT_COMPILE_COMMANDS=ON`

+1. We have the same usage and issue. Commands generated by ninja have relative paths for the dependencies in the exported `build_commands.json` and there's no way to toggle the path format.

**Comment by bradking** [comment]
> `cmake ../ -GNinja`

When the build tree is placed inside the source tree, CMake chooses to use relative paths starting in `../` to reference source files from `build.ninja`.  If the build tree is placed *outside* the source tree, CMake will reference files in the source tree by absolute path.  [CMake Issue 13894](https://gitlab.kitware.com/cmake/cmake/-/issues/13894) proposes changing to absolute paths in both cases.

**Comment by kobalicek** [comment]
@bradking It would be really nice to get this fixed somehow. To be honest I'm just an end user of both CMake and Ninja. But because of this issue I started using unix makefiles generator again.

I don't really expect the PR to be merged, so maybe CMake defaulting to absolute paths or giving users some option to do so?

**Comment by JBakamovic** [comment]
@bradking Above was just an example of command. In reality I am using out-of-source builds and the issue with relative paths persists.

Repro:
> $ cd /tmp 
> $ git clone https://github.com/mysql/mysql-server.git
> $ mkdir build_ninja
> $ cd build_ninja
> $ cmake ../mysql-server -GNinja -DCMAKE_BUILD_TYPE=Debug -DDOWNLOAD_BOOST=1 -DWITH_BOOST=../ -DWITH_UNIT_TESTS=ON -DCMAKE_EXPORT_COMPILE_COMMANDS=ON
> $ cd /tmp
> $ mkdir build_makefile
> $ cd build_makefile
> $ cmake ../mysql-server -G "Unix Makefiles" -DCMAKE_BUILD_TYPE=Debug -DDOWNLOAD_BOOST=1 -DWITH_BOOST=../ -DWITH_UNIT_TESTS=ON -DCMAKE_EXPORT_COMPILE_COMMANDS=ON
> $ tail /tmp/build_ninja/compile_commands.json
```
"directory": "/tmp/build_ninja",
  "command": "/usr/lib64/ccache/c++ -DHAVE_CONFIG_H -DHAVE_TLSv13 -DLZ4_DISABLE_DEPRECATE_WARNINGS -DMYSQL_ROUTER_LOG_DOMAIN=\\\"routing\\\" -DRAPIDJSON_NO_SIZETYPEDEFINE -DRAPIDJSON_SCHEMA_USE_INTERNALREGEX=0 -DRAPIDJSON_SCHEMA_USE_STDREGEX=1 -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_USE_MATH_DEFINES -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Drouting_EXPORTS -I. -Iinclude -I/tmp/mysql-server -I/tmp/mysql-server/include -Irouter -Irouter/include -I/tmp/mysql-server/router/src/metadata_cache/include -I/tmp/mysql-server/router/src/routing/include -Irouter/src/routing/include -I/tmp/mysql-server/router/src/harness/src/../include -Irouter/src/harness/src/src_SHARED -I/tmp/mysql-server/router/src/mysql_protocol/include -I/tmp/mysql-server/router/src/router/src/../include -Irouter/src/harness/src/../include -I/tmp/mysql-server/router/src/io/src/../include -Irouter/src/io/src/../include -I/tmp/mysql-server/router/src/harness/include -isystem /tmp/mysql-server/extra/protobuf/protobuf-3.11.4/src -isystem /tmp/mysql-server/extra/rapidjson/include -isystem /tmp/mysql-server/extra/lz4 -isystem /tmp/mysql-server/extra/libedit/libedit-20191231-3.1/src/editline -isystem /tmp/mysql-server/extra/zstd/lib -isystem extra/zlib -isystem /tmp/mysql-server/extra/zlib -isystem plugin/x/generated/protobuf_lite -std=c++14 -fno-omit-frame-pointer -ftls-model=initial-exec  -Wall -Wextra -Wformat-security -Wvla -Wundef -Wmissing-format-attribute -Woverloaded-virtual -Wcast-qual -Wimplicit-fallthrough=2 -Wstringop-truncation -Wsuggest-override -Wmissing-include-dirs -Wlogical-op -Werror -DSAFE_MUTEX -DENABLED_DEBUG_SYNC -g -fPIC -o router/src/routing/CMakeFiles/routing.dir/src/x_protocol_splicer.cc.o -c /tmp/mysql-server/router/src/routing/src/x_protocol_splicer.cc",
  "file": "/tmp/mysql-server/router/src/routing/src/x_protocol_splicer.cc"
},
{
  "directory": "/tmp/build_ninja",
  "command": "/usr/lib64/ccache/c++ -DHAVE_CONFIG_H -DHAVE_TLSv13 -DLZ4_DISABLE_DEPRECATE_WARNINGS -DMYSQL_ROUTER_LOG_DOMAIN=\\\"routing\\\" -DRAPIDJSON_NO_SIZETYPEDEFINE -DRAPIDJSON_SCHEMA_USE_INTERNALREGEX=0 -DRAPIDJSON_SCHEMA_USE_STDREGEX=1 -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_USE_MATH_DEFINES -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Drouting_EXPORTS -I. -Iinclude -I/tmp/mysql-server -I/tmp/mysql-server/include -Irouter -Irouter/include -I/tmp/mysql-server/router/src/metadata_cache/include -I/tmp/mysql-server/router/src/routing/include -Irouter/src/routing/include -I/tmp/mysql-server/router/src/harness/src/../include -Irouter/src/harness/src/src_SHARED -I/tmp/mysql-server/router/src/mysql_protocol/include -I/tmp/mysql-server/router/src/router/src/../include -Irouter/src/harness/src/../include -I/tmp/mysql-server/router/src/io/src/../include -Irouter/src/io/src/../include -I/tmp/mysql-server/router/src/harness/include -isystem /tmp/mysql-server/extra/protobuf/protobuf-3.11.4/src -isystem /tmp/mysql-server/extra/rapidjson/include -isystem /tmp/mysql-server/extra/lz4 -isystem /tmp/mysql-server/extra/libedit/libedit-20191231-3.1/src/editline -isystem /tmp/mysql-server/extra/zstd/lib -isystem extra/zlib -isystem /tmp/mysql-server/extra/zlib -isystem plugin/x/generated/protobuf_lite -std=c++14 -fno-omit-frame-pointer -ftls-model=initial-exec  -Wall -Wextra -Wformat-security -Wvla -Wundef -Wmissing-format-attribute -Woverloaded-virtual -Wcast-qual -Wimplicit-fallthrough=2 -Wstringop-truncation -Wsuggest-override -Wmissing-include-dirs -Wlogical-op -Werror -DSAFE_MUTEX -DENABLED_DEBUG_SYNC -g -fPIC -o router/src/routing/CMakeFiles/routing.dir/src/destination_ssl_context.cc.o -c /tmp/mysql-server/router/src/routing/src/destination_ssl_context.cc",
  "file": "/tmp/mysql-server/router/src/routing/src/destination_ssl_context.cc"
}
]
```
> $ tail /tmp/build_makefile/compile_commands.json
```
  "directory": "/tmp/build_makefile/router/src/routing",
  "command": "/usr/lib64/ccache/c++ -DHAVE_CONFIG_H -DHAVE_TLSv13 -DLZ4_DISABLE_DEPRECATE_WARNINGS -DMYSQL_ROUTER_LOG_DOMAIN=\\\"routing\\\" -DRAPIDJSON_NO_SIZETYPEDEFINE -DRAPIDJSON_SCHEMA_USE_INTERNALREGEX=0 -DRAPIDJSON_SCHEMA_USE_STDREGEX=1 -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_USE_MATH_DEFINES -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Drouting_EXPORTS -I/tmp/build_makefile -I/tmp/build_makefile/include -I/tmp/mysql-server -I/tmp/mysql-server/include -I/tmp/build_makefile/router -I/tmp/build_makefile/router/include -I/tmp/mysql-server/router/src/metadata_cache/include -I/tmp/mysql-server/router/src/routing/include -I/tmp/build_makefile/router/src/routing/include -I/tmp/mysql-server/router/src/harness/src/../include -I/tmp/build_makefile/router/src/harness/src/src_SHARED -I/tmp/mysql-server/router/src/mysql_protocol/include -I/tmp/mysql-server/router/src/router/src/../include -I/tmp/build_makefile/router/src/harness/src/../include -I/tmp/mysql-server/router/src/io/src/../include -I/tmp/build_makefile/router/src/io/src/../include -I/tmp/mysql-server/router/src/harness/include -isystem /tmp/mysql-server/extra/protobuf/protobuf-3.11.4/src -isystem /tmp/mysql-server/extra/rapidjson/include -isystem /tmp/mysql-server/extra/lz4 -isystem /tmp/mysql-server/extra/libedit/libedit-20191231-3.1/src/editline -isystem /tmp/mysql-server/extra/zstd/lib -isystem /tmp/build_makefile/extra/zlib -isystem /tmp/mysql-server/extra/zlib -isystem /tmp/build_makefile/plugin/x/generated/protobuf_lite -std=c++14 -fno-omit-frame-pointer -ftls-model=initial-exec  -Wall -Wextra -Wformat-security -Wvla -Wundef -Wmissing-format-attribute -Woverloaded-virtual -Wcast-qual -Wimplicit-fallthrough=2 -Wstringop-truncation -Wsuggest-override -Wmissing-include-dirs -Wlogical-op -Werror -DSAFE_MUTEX -DENABLED_DEBUG_SYNC -g -fPIC -o CMakeFiles/routing.dir/src/x_protocol_splicer.cc.o -c /tmp/mysql-server/router/src/routing/src/x_protocol_splicer.cc",
  "file": "/tmp/mysql-server/router/src/routing/src/x_protocol_splicer.cc"
},
{
  "directory": "/tmp/build_makefile/router/src/routing",
  "command": "/usr/lib64/ccache/c++ -DHAVE_CONFIG_H -DHAVE_TLSv13 -DLZ4_DISABLE_DEPRECATE_WARNINGS -DMYSQL_ROUTER_LOG_DOMAIN=\\\"routing\\\" -DRAPIDJSON_NO_SIZETYPEDEFINE -DRAPIDJSON_SCHEMA_USE_INTERNALREGEX=0 -DRAPIDJSON_SCHEMA_USE_STDREGEX=1 -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_USE_MATH_DEFINES -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Drouting_EXPORTS -I/tmp/build_makefile -I/tmp/build_makefile/include -I/tmp/mysql-server -I/tmp/mysql-server/include -I/tmp/build_makefile/router -I/tmp/build_makefile/router/include -I/tmp/mysql-server/router/src/metadata_cache/include -I/tmp/mysql-server/router/src/routing/include -I/tmp/build_makefile/router/src/routing/include -I/tmp/mysql-server/router/src/harness/src/../include -I/tmp/build_makefile/router/src/harness/src/src_SHARED -I/tmp/mysql-server/router/src/mysql_protocol/include -I/tmp/mysql-server/router/src/router/src/../include -I/tmp/build_makefile/router/src/harness/src/../include -I/tmp/mysql-server/router/src/io/src/../include -I/tmp/build_makefile/router/src/io/src/../include -I/tmp/mysql-server/router/src/harness/include -isystem /tmp/mysql-server/extra/protobuf/protobuf-3.11.4/src -isystem /tmp/mysql-server/extra/rapidjson/include -isystem /tmp/mysql-server/extra/lz4 -isystem /tmp/mysql-server/extra/libedit/libedit-20191231-3.1/src/editline -isystem /tmp/mysql-server/extra/zstd/lib -isystem /tmp/build_makefile/extra/zlib -isystem /tmp/mysql-server/extra/zlib -isystem /tmp/build_makefile/plugin/x/generated/protobuf_lite -std=c++14 -fno-omit-frame-pointer -ftls-model=initial-exec  -Wall -Wextra -Wformat-security -Wvla -Wundef -Wmissing-format-attribute -Woverloaded-virtual -Wcast-qual -Wimplicit-fallthrough=2 -Wstringop-truncation -Wsuggest-override -Wmissing-include-dirs -Wlogical-op -Werror -DSAFE_MUTEX -DENABLED_DEBUG_SYNC -g -fPIC -o CMakeFiles/routing.dir/src/destination_ssl_context.cc.o -c /tmp/mysql-server/router/src/routing/src/destination_ssl_context.cc",
  "file": "/tmp/mysql-server/router/src/routing/src/destination_ssl_context.cc"
}
]
```

First obvious difference is in `-I. -Iinclude` vs `-I/tmp/build_makefile -I/tmp/build_makefile/include`  but there're plenty of other relative paths which are not present in Makefile compile_commands.json. E.g. `-Irouter`, `-Irouter/include`, `-Irouter/src/routing/include`, `-isystem extra/zlib`, `-isystem plugin/x/generated/protobuf_lite`, etc. etc.

**Comment by bradking** [comment]
@JBakamovic thanks for the example.  However, CMake's behavior would be best discussed over in CMake's issue tracker.  See [CMake Issue 13894](https://gitlab.kitware.com/cmake/cmake/-/issues/13894).  `compile_commands.json` is generated using the same command lines that are put in `build.ninja`.  They use relative paths in many cases because that is what Ninja conventions prefer.  CMake can probably switch to absolute paths for more cases than it has now, but paths inside the build tree probably need to remain relative unless Ninja addresses the concern posted in [this comment above](https://github.com/ninja-build/ninja/issues/1251#issuecomment-793010993).

**Comment by dvolosnykh** [comment]
In the same way as @kobalicek, I am the end user of CMake using Ninja as a backend. I also had to switch our usage of CMake to `Unix Makefiles` generator due to this issue.

What really disappoints me is a 4-year holy war around relative vs absolute paths resulting in nothing. I believe the tool shall give choices.

@bradking I am really thankful for your efforts in pushing this issue for such a long time.

**Comment by bradking** [comment]
CMake 3.21 includes a fix to [CMake Issue 13894](https://gitlab.kitware.com/cmake/cmake/-/issues/13894) that does not require changes to Ninja.  Instead CMake uses a workaround to this issue by expressing outputs as both relative and absolute paths.  It causes Ninja to be less efficient by `stat`ing each output file twice, but at least we can use absolute paths.

**Comment by GrandChris** [comment]
@bradking That sounds great, I'd also like to thank you for your efforts on this issue. I've been following this thread for months now, personally I've already given up that there will ever be a solution.

**Comment by GrandChris** [comment]
@bradking I've been eager and tried the CMake 3.21.0 version which uses absolute paths and is great. What I noticed is that dependencies no longer work with generated headers.
Specifically: I create header files from Json files with "add_custom_target" via CMake. The header files are included in cpp files, which means that Ninja must discover these headers and include them in the dependency tree. Apparently this only works with relative paths.
For me, this means that if I made some changes in a Json file, I have to compile my project twice in order for those changes to be applied to the executable. For now I rather deal with this than with relative paths, but this is still not perfect.

**Comment by bradking** [comment]
@GrandChris please open an issue in CMake's issue tracker for that.

**Comment by dvolosnykh** [comment]
@bradking @GrandChris Has the issue with generated headers been filed in CMake's issue tracker? Could not find anything related.

**Comment by bradking** [comment]
@dvolosnykh it is [CMake Issue 22454](https://gitlab.kitware.com/cmake/cmake/-/issues/22454), and turned out not to be a regression or a problem caused by the switch to absolute paths.

**Comment by motoz** [comment]
I'm having trouble with our custom generator where i try to support both in tree and completely out of tree builds. To me it seems that I have no control over when gcc starts outputting absolute paths in the .d files, which then breaks incremental builds badly. I have taken great pains to use relative paths everywhere but despite that plain old gcc (on windows, but still) starts to spit out absolute paths seemingly when it finds a header using a relative paths the 'touches root' or something like that. I'd like to be able to point the generator with e.g. --builddir=/temp/testbuild while running the configurator (and ninja) from the source tree. The problem is that I generate headers during the build that I naturally place in the build directory (which then is referenced by e.g. -I../../temp/testbuild, i.e. relatively), but despite that gcc insists on switching to an absolute path for headers found from that build directory. If I could work out the exact logic how gcc generates the path (well I guess I could look at the source...) I suppose I should be able to work around it in the generator, but it would be so much easier if ninja could handle it.

**Comment by GrandChris** [comment]
Hello @motoz , relative paths are fine, if all paths start at the root of your project directory. If this is not possible, e.g. because some tool insists of starting all paths from a build directory, I highly recommend using absolute paths for everything. Otherwise you will keep fighting all sorts of issues with your paths.

**Comment by jhasse** [comment]
#2667 brought this issue up to my attention again. The discussion has been long and we have mixed multiple issues in here. So I want to separate them into several smaller issues so that we can fix them.

I think a very helpful addition to ninja would a mode where it checks that everything is correct and if not warns about it. This would be disabled by default because it comes with a performance cost (because of all the real pathing). That's #2719.

Other issues about abs/rel paths: #2720 and #2283. Feel free to create more :)

**Comment by bradking** [comment]
@jhasse thanks.  The primary issue here is reconciling depfile-loaded absolute paths with relative build manifest paths, and vice-versa.  Do you consider #2720 an appropriate place for that, or should I open another issue?

**Comment by jhasse** [comment]
As #2720 isn't about depfiles, please open a new issue :)

**Comment by bradking** [comment]
#2721 replaces this issue.
